node/src/components/consensus/highway_core/state.rs

```diff
@@ -566,8 +575,54 @@ impl<C: Context> State<C> {
                 return Err(VoteError::ValueAfterTerminalBlock);
             }
         }
-        // TODO: Validate against LNC.
-        Ok(())
+        match self.validate_lnc(wvote) {
+            Ok(_) => Ok(()),
+            Err(LncError::MissingEndorsement(vhash)) => {
+                panic!("Protocol state is missing endorsement for {:?}", &vhash)
+            }
+            Err(LncError::NaiveCitation(vidx)) => Err(VoteError::LncNaiveCitation(vidx)),
+        }
+    }
+
+    /// Validates whether `wvote` violates the LNC rule.
+    /// Returns the first vote's hash that should have been endorsed but wasn't.
+    ///
+    /// Vote violates LNC rule if it cites naively an equivocation.
+    /// If it cites equivocator then it needs to endorse votes that cite equivocating votes.
+    fn validate_lnc(&self, wvote: &WireVote<C>) -> Result<(), LncError<C>> {
+        let mut need_endorsement = wvote.panorama.need_endorsements(self);
```

Comment:
But not all of those are relevant to the LNC for `wvote`, are they?
The LNC only talks about naive citations by units with the same creator as `wvote`.
Also, the endorsement tracker doesn't seem to check which of the returned citations actually are naive: Some of these units may be citing an equivocator non-naively, because _they_ claim that some unit in between was endorsed.

Reply:
And even among those that are naive, not all are relevant:
Some of them may cite an equivocator, but _before_ their fork, i.e. the cited unit might not actually be part of the equivocation.

Reply:
> The LNC only talks about naive citations by units with the same creator as `wvote`.

My interpretation was that it can also be `wvote` that cites equivocations:

> A unit _u_ by validator _V_ is considered incorrect if there is an equivocation (_v1_, _v2_) (created by some validator _W_) and two units _u1, u2 ≤ u_ created by _V_ such that _u1 cites naively v1_ and _u2 cites naively v2_.

So `wvote` can be _u1_ and _u2_.

Reply:
Yes, but even if _u1_ and _u2_ are not both equal to _u_, we are only interested in them if they are by _V_.
The endorsement tracker collects all citations, even by votes that are not created by _V_.

Reply:
Because someone else might have created unit _w_ that is endorsed and _u1 > w > v1_ ? In this case _u1_ cites _V_ non-naively but endorsed vote is not by _V_.

Reply:
Right, but then we don't need _u1_ **and** _w_ to be endorsed; just _u1_ **or** _w_.

Reply:
This implementation should return _w_ only – i.e. the first vote that cites an equivocator. Although, it breaks b/c of what you wrote earlier: some votes may cite an equivocator but _before_ the fork.